subprojects/internal-instrumentation-processor/src/main/java/org/gradle/internal/instrumentation/processor/AbstractInstrumentationProcessor.java

```diff
@@ -73,11 +79,35 @@ private Set<Class<? extends Annotation>> getSupportedAnnotations() {
 
     @Override
     public boolean process(Set<? extends TypeElement> annotations, RoundEnvironment roundEnv) {
-        Set<? extends Element> annotatedTypes = getSupportedAnnotations().stream().flatMap(annotation -> roundEnv.getElementsAnnotatedWith(annotation).stream()).collect(Collectors.toSet());
+        Set<? extends Element> annotatedTypes = getSupportedAnnotations().stream()
+            .flatMap(annotation -> roundEnv.getElementsAnnotatedWith(annotation).stream())
+            .collect(Collectors.toSet());
+        annotatedTypes = findActualTypesToVisit(annotatedTypes);
         collectAndProcessRequests(annotatedTypes);
         return true;
     }
 
+    private Set<? extends Element> findActualTypesToVisit(Set<? extends Element> annotatedTypes) {
+        return annotatedTypes.stream()
+            .flatMap(it -> findActualTypesToVisit(it).stream())
+            .collect(Collectors.toSet());
```

Comment:
Note that `Collectors.toSet()` gives non-deterministic ordering of the results. Please consider collecting the results to a list or an ordered set using `toCollection()` instead.

Reply:
Another option would be to do away with collectors, and rely on streams instead:

```patch
diff --git a/subprojects/internal-instrumentation-processor/src/main/java/org/gradle/internal/instrumentation/processor/AbstractInstrumentationProcessor.java b/subprojects/internal-instrumentation-processor/src/main/java/org/gradle/internal/instrumentation/processor/AbstractInstrumentationProcessor.java
index 50dfec905d4..e7a5e3a0e9d 100644
--- a/subprojects/internal-instrumentation-processor/src/main/java/org/gradle/internal/instrumentation/processor/AbstractInstrumentationProcessor.java
+++ b/subprojects/internal-instrumentation-processor/src/main/java/org/gradle/internal/instrumentation/processor/AbstractInstrumentationProcessor.java
@@ -79,20 +79,13 @@ private Set<Class<? extends Annotation>> getSupportedAnnotations() {
 
     @Override
     public boolean process(Set<? extends TypeElement> annotations, RoundEnvironment roundEnv) {
-        Set<? extends Element> annotatedTypes = getSupportedAnnotations().stream()
+        Stream<? extends Element> annotatedTypes = getSupportedAnnotations().stream()
             .flatMap(annotation -> roundEnv.getElementsAnnotatedWith(annotation).stream())
-            .collect(Collectors.toSet());
-        annotatedTypes = findActualTypesToVisit(annotatedTypes);
+            .flatMap(it -> findActualTypesToVisit(it).stream());
         collectAndProcessRequests(annotatedTypes);
         return true;
     }
 
-    private Set<? extends Element> findActualTypesToVisit(Set<? extends Element> annotatedTypes) {
-        return annotatedTypes.stream()
-            .flatMap(it -> findActualTypesToVisit(it).stream())
-            .collect(Collectors.toSet());
-    }
-
     private Set<Element> findActualTypesToVisit(Element typeElement) {
         Optional<? extends AnnotationMirror> annotationMirror = AnnotationUtils.findAnnotationMirror(typeElement, VisitForInstrumentation.class);
         if (!annotationMirror.isPresent()) {
@@ -112,10 +105,10 @@ private <T extends InstrumentationProcessorExtension> Collection<T> getExtension
         return Cast.uncheckedCast(getExtensions().stream().filter(type::isInstance).collect(Collectors.toList()));
     }
 
-    private void collectAndProcessRequests(Collection<? extends Element> annotatedElements) {
+    private void collectAndProcessRequests(Stream<? extends Element> annotatedElements) {
         Collection<AnnotatedMethodReaderExtension> readers = getExtensionsByType(AnnotatedMethodReaderExtension.class);
 
-        List<ExecutableElement> allMethodElementsInAnnotatedClasses = getExecutableElementsFromAnnotatedElements(annotatedElements);
+        Stream<ExecutableElement> allMethodElementsInAnnotatedClasses = getExecutableElementsFromAnnotatedElements(annotatedElements);
 
         Map<ExecutableElement, List<CallInterceptionRequestReader.Result.InvalidRequest>> errors = new LinkedHashMap<>();
         List<CallInterceptionRequestReader.Result.Success> successResults = new ArrayList<>();
@@ -133,8 +126,8 @@ private void collectAndProcessRequests(Collection<? extends Element> annotatedEl
     }
 
     @Nonnull
-    private static List<ExecutableElement> getExecutableElementsFromAnnotatedElements(Collection<? extends Element> annotatedClassElements) {
-        return annotatedClassElements.stream()
+    private static Stream<ExecutableElement> getExecutableElementsFromAnnotatedElements(Stream<? extends Element> annotatedClassElements) {
+        return annotatedClassElements
             .flatMap(element -> element.getKind() == ElementKind.METHOD ? Stream.of(element) : element.getEnclosedElements().stream())
             .filter(it -> it.getKind() == ElementKind.METHOD)
             .map(it -> (ExecutableElement) it)
@@ -142,11 +135,11 @@ private static List<ExecutableElement> getExecutableElementsFromAnnotatedElement
             // The order in which the executable elements are listed should be the order in which they appear in the code but
             // we take an extra measure of care here and ensure the ordering between all elements.
             .sorted(Comparator.comparing(AbstractInstrumentationProcessor::elementQualifiedName))
-            .collect(Collectors.toList());
+            .distinct();
     }
 
-    private static void readRequests(Collection<AnnotatedMethodReaderExtension> readers, List<ExecutableElement> allMethodElementsInAnnotatedClasses, Map<ExecutableElement, List<CallInterceptionRequestReader.Result.InvalidRequest>> errors, List<CallInterceptionRequestReader.Result.Success> successResults) {
-        for (ExecutableElement methodElement : allMethodElementsInAnnotatedClasses) {
+    private static void readRequests(Collection<AnnotatedMethodReaderExtension> readers, Stream<ExecutableElement> allMethodElementsInAnnotatedClasses, Map<ExecutableElement, List<CallInterceptionRequestReader.Result.InvalidRequest>> errors, List<CallInterceptionRequestReader.Result.Success> successResults) {
+        allMethodElementsInAnnotatedClasses.forEach(methodElement -> {
             for (AnnotatedMethodReaderExtension reader : readers) {
                 Collection<CallInterceptionRequestReader.Result> readerResults = reader.readRequest(methodElement);
                 for (CallInterceptionRequestReader.Result readerResult : readerResults) {
@@ -157,7 +150,7 @@ private static void readRequests(Collection<AnnotatedMethodReaderExtension> read
                     }
                 }
             }
-        }
+        });
     }
 
     @Nonnull
```

Reply:
Replaced with streams as suggested, since it simplifies some cases (especially we can get rid `findActualTypesToVisit`. I left List for `getExecutableElementsFromAnnotatedElements`, since after that we don't do any operation on streams anymore